src/gc/proxy.d

```diff
@@ -28,32 +28,67 @@ private
     extern (C) void thread_init();
     extern (C) void thread_term();
 
+    import core.internal.spinlock;
+    static SpinLock instanceLock;
+
     __gshared GC instance;
     __gshared GC proxiedGC; // used to iterate roots of Windows DLLs
-
 }
 
+alias GCInitNoThrow = void function() nothrow @nogc;
 
 extern (C)
 {
 
     void gc_init()
     {
-        config.initialize();
-        ManualGC.initialize(instance);
-        ConservativeGC.initialize(instance);
-        if (instance is null)
-        {
-            import core.stdc.stdio : fprintf, stderr;
-            import core.stdc.stdlib : exit;
+        import core.atomic : atomicLoad, atomicStore;
+
+        auto pinstance = cast(shared(GC)*) &instance;
 
-            fprintf(stderr, "No GC was initialized, please recheck the name of the selected GC ('%.*s').\n", cast(int)config.gc.length, config.gc.ptr);
-            exit(1);
+        if (atomicLoad(*pinstance) is null)
+        {
+            instanceLock.lock();
+            scope(exit) instanceLock.unlock();
+            {
+                if (atomicLoad(*pinstance) is null)
+                {
+                    GC auxInstance;
+                    config.initialize();
+                    ManualGC.initialize(auxInstance);
+                    ConservativeGC.initialize(auxInstance);
+
+                    if (auxInstance is null)
+                    {
+                        import core.stdc.stdio : fprintf, stderr;
+                        import core.stdc.stdlib : exit;
+
+                        fprintf(stderr, "No GC was initialized, please recheck the name of the selected GC ('%.*s').\n", cast(int)config.gc.length, config.gc.ptr);
+                        instanceLock.unlock();
```

Comment:
You've unlocked twice!  (`scope(exit) instanceLock.unlock();` above)

Reply:
yes please delete

Reply:
For some reason, if unlock is not called before ```exit(1)```, this test hangs: ```druntime/test/exceptions/src/unknown_gc.d```

Reply:
uh oh, that may be a bug with D's scope(exit) implementation (anyone correct me if I'm wrong).  Given this, I would suggest not using `scope(exit)` here and instead insert `instancelock.unlock()` at every "exit point".  There should just be 2, just before calling `exit` and at the end of the block.

Reply:
FWIW we have a nice language feature for exactly this:

```d
synchronized (instanceLock)
{
   // code here
}
```

Reply:
If `scope(exit)` doesn't work with the `exit` function, my guess is that this might suffer from the same bug.  @somzzz can you try this out and confirm?

Reply:
Wait, `exit()` doesn't return, so it can't exit the scope. This seems like the code is correct to me...

`scope(exit)` only works if the scope is exited, via stack unwinding or normal execution. Calling exit function is neither.

Reply:
keep scope(exit) and also unlock explicitly before calling exit thx